### PR TITLE
fix(query): Escape Special chars

### PR DESCRIPTION
**Pull request by astalker**
Solr throws a parse exception when it encounters special characters that aren't escaped.

https://lucene.apache.org/core/2_9_4/queryparsersyntax.html#Escaping Special Characters

There is a Solr Utils package but a straight forward str_replace seems fine enough.

I've used a slightly modified replace (omitting the * and space) from http://e-mats.org/2010/01/escaping-characters-in-a-solr-query-solr-url/

Might be better in the Solr package somewhere.

### CODE REVIEW

**Comment by shieldo** [comment]
Can you explain what you mean by a "special character" - do you mean non-ASCII or something like that? Any specific examples? Any possible unit tests?

**Comment by astalker** [comment]
Yeah I can write unit tests. Just wanted to check I had the correct approach first.

Maybe click the links? ;)

File: Terms/SolrRegexTermsService.php [hunk 1]
```diff
@@ -42,4 +42,13 @@ public function fetchTerms(SimpleQueryInterface $query)
 
         return new SolrTermsResult($resultSet);
     }
+
+    private function escapeSolrValue($string)
+    {
+        $match = ['\\', '+', '-', '&', '|', '!', '(', ')', '{', '}', '[', ']', '^', '~', '?', ':', '"', ';'];
+        $replace = ['\\\\', '\\+', '\\-', '\\&', '\\|', '\\!', '\\(', '\\)', '\\{', '\\}', '\\[', '\\]', '\\^', '\\~', '\\?', '\\:', '\\"', '\\;'];
+        $string = str_replace($match, $replace, $string);
```
Comment (shieldo) [inline Terms/SolrRegexTermsService.php]: Just return the expression result?

**Comment by shieldo** [comment]
Ah so you mean characters that have meaning in Lucene expressions?

iirc we already escape these in the normal query service... looking...

**Comment by shieldo** [comment]
Think we already make use of the class `Solarium\Core\Query\Helper` from Solarium here?

**Comment by astalker** [comment]
Updated to use a slight variation of the Helper method

**Comment by calumbrodie** [comment]
@astalker Isn't this a bug against solarium? I find it hard to believe this is required fix in our client library. What is the failing test case?

**Comment by shieldo** [comment]
Actually looking at this @astalker - are you sure this is to do with Lucene expressions and not the fact that you are setting a regex field on the terms query and that characters like `*` are significant in regexes?

Given this, it's certainly an issue for Needle and not Solarium - Needle shouldn't be passing strings verbatim as regexes.

**Comment by astalker** [comment]
@calumbrodie I tried the helper method but it also escapes `*` which breaks the query. So I replicated the method and removed `*`. Is it Solarium that constructs the query? i.e. adds the `*` to the term? Maybe it should escape when creating the actual search term? I didn't really see where that was.

A functional test case is to search a term with one of the chars like `[`. 

@shieldo The `*` is fine, its the other characters that need escaped.

File: Terms/SolrRegexTermsService.php [hunk 2]
```diff
@@ -24,10 +24,9 @@ public function fetchTerms(SimpleQueryInterface $query)
             return new EmptyTermsResult();
         }
 
-        $suggestQuery
-            ->setFields($field)
-            ->setRegex($query->getSearchTerm())
-            ->setSort('count');
+        $suggestQuery->setFields($field);
+        $suggestQuery->setRegex($this->escapeTerm($query->getSearchTerm()));
```
Comment (shieldo) [inline Terms/SolrRegexTermsService.php]: This field expects a regular expression value, *not* a Lucene expression.

https://cwiki.apache.org/confluence/display/solr/The+Terms+Component (under terms.regex)

To escape, try using `preg_quote()` or similar.

Comment (astalker) [inline Terms/SolrRegexTermsService.php]: Isn't this doing the same as the `escapeTerm` helper method?

This is where the method refers to regarding special characters though the link is dead
http://lucene.apache.org/core/queryparsersyntax.html#Escaping%20Special%20Characters

`preg_quote` doesn't work for the same reason.

Comment (astalker) [inline Terms/SolrRegexTermsService.php]: You end up with a `terms.regex` of `\.\*blue\.\*`

Comment (shieldo) [inline Terms/SolrRegexTermsService.php]: Is that bad?

Comment (astalker) [inline Terms/SolrRegexTermsService.php]: Yeah, doesn't work as in the query fails. Otherwise I'd have just used the helper method.

**Pull request closed without merging**